lib/rules/no-misleading-unicode-character.ts

```diff
@@ -0,0 +1,276 @@
+        return "Surrogate"
+    }
+    return null
+}
+
+/** Returns the last grapheme of the quantified element. */
+function getGraphemeBeforeQuant(quant: Quantifier): string {
+    const alt = quant.parent
+
+    // find the start index of the first character left of
+    // this quantifier
+    let start = quant.start
+    for (let i = alt.elements.indexOf(quant) - 1; i >= 0; i--) {
+        const e = alt.elements[i]
+        if (e.type === "Character" && !isEscapeSequence(e.raw)) {
+            start = e.start
+        } else {
+            break
+        }
+    }
+
+    const before = alt.raw.slice(
+        start - alt.start,
+        quant.element.end - alt.start,
+    )
+
+    const graphemes = splitter.splitGraphemes(before)
+    const grapheme = graphemes[graphemes.length - 1]
+
+    return grapheme
+}
+
+interface GraphemeProblem {
+    readonly grapheme: string
+    readonly problem: Problem
+    readonly start: number
+    readonly end: number
+    /** A sorted list of all unique elements that overlap with this grapheme */
+    readonly elements: CharacterClassElement[]
+}
+
+/** Returns all grapheme problem in the given character class. */
+function getGraphemeProblems(
+    cc: CharacterClass,
+    flags: ReadonlyFlags,
+): GraphemeProblem[] {
+    let offset = cc.negate ? 2 : 1
+
+    const graphemes = splitter.splitGraphemes(cc.raw.slice(offset, -1))
+    const problems: GraphemeProblem[] = []
+
+    for (const grapheme of graphemes) {
+        const problem = getProblem(grapheme, flags)
+        if (problem !== null) {
+            const start = offset + cc.start
+            const end = start + grapheme.length
+
+            problems.push({
+                grapheme,
+                problem,
+                start,
+                end,
+                elements: cc.elements.filter(
+                    (e) => e.start < end && e.end > start,
+                ),
+            })
+        }
+        offset += grapheme.length
+    }
+
+    return problems
+}
+
+/** Returns a fix for the given problems (if possible). */
+function getGraphemeProblemsFix(
+    problems: readonly GraphemeProblem[],
+    cc: CharacterClass,
+): string | null {
+    if (cc.negate) {
+        // we can't fix a negated character class
+        return null
+    }
+
+    if (
+        !problems.every(
+            (p) =>
+                p.start === p.elements[0].start &&
+                p.end === p.elements[p.elements.length - 1].end,
+        )
+    ) {
+        // the graphemes don't line up with character class elements
+        return null
+    }
+
+    // The prefix of graphemes
+    const prefix = problems
+        .map((p) => p.grapheme)
+        .sort((a, b) => b.length - a.length)
+        .join("|")
+
+    // The rest of the character class
+    let ccRaw = cc.raw
+    for (let i = problems.length - 1; i >= 0; i--) {
+        const { start, end } = problems[i]
+        ccRaw = ccRaw.slice(0, start) + ccRaw.slice(end)
```

Comment:
Maybe we need to do the following:

```suggestion
        ccRaw = ccRaw.slice(0, start - cc.start) + ccRaw.slice(end - cc.start)
```

I found that it couldn't be autofixed well in the following cases:

```js
/[竈門禰󠄀豆子]|[煉󠄁獄杏寿郎]/
```